website/blog/2019-08-19-tablescan-structs.md

```diff
@@ -0,0 +1,196 @@
+```
+| Version  | Wall time (seconds) | CPU time (seconds) | Baseline CPU / Aria CPU |
+| -------- | ------------------- | ------------------ | ----------------------- |
+| Aria     | 4                   | 35                 | 1.0                     |
+| Baseline | 7                   | 80                 | 2.28                    |
+
+```sql
+SELECT COUNT(*)
+FROM exportlineitem
+WHERE shipment.partkey BETWEEN 1000000 AND 2000000 AND shipment.suppkey BETWEEN 100000 AND 200000 AND shipment.extendedprice > 0 ;
+```
+| Version  | Wall time (seconds) | CPU time (seconds) | Baseline CPU / Aria CPU |
+| -------- | ------------------- | ------------------ | ----------------------- |
+| Aria     | 4                   | 35                 | 1.0                     |
+| Baseline | 16                  | 227                | 6.5                     |
+
+We notice that for Aria it makes no difference whether the filtered columns are top level or in a non-null struct. We also note that none of the columns are materialize by Aria, since these are only filtered on, but they are materialized by baseline Presto.
+
+```sql
+SELECT COUNT(*), SUM(shipment.extendedprice)
+FROM exportlineitem
+WHERE shipment.partkey BETWEEN 1000000 AND 2000000 AND shipment.suppkey BETWEEN 100000 AND 200000;
+```
+| Version  | Wall time (seconds) | CPU time (seconds) | Baseline CPU / Aria CPU |
+| -------- | ------------------- | ------------------ | ----------------------- |
+| Aria     | 4                   | 34                 | 1.0                     |
+| Baseline | 18                  | 253                | 7.44                    |
+
+Now, instead of having a filter that is always true, we retrieve the value of `extendedprice` and materialize a struct. This takes only marginally longer because the struct only has the `extendedprice` column filled in. The struct needs to be materialized because it is an input to the SUM aggregate. We could of course, since we only access fields of the struct and not the whole struct, elide materializing the struct and only materialize the component columns. But the gain of this last optimization would not be very large.
+
+```sql
+SELECT COUNT(*), SUM(shipment.extendedprice), COUNT(export.license.filing_no)
+FROM exportlineitem
+WHERE shipment.suppkey BETWEEN 200000 AND 400000 AND shipment.quantity < 10;
+```
+| Version  | Wall time (seconds) | CPU time (seconds) | Baseline CPU / Aria CPU |
+| -------- | ------------------- | ------------------ | ----------------------- |
+| Aria     | 10                  | 58                 | 1.0                     |
+| Baseline | 30                  | 330                | 5.68                    |
+
+Here we add a second struct to the mix. We filter on members of one struct and return a field, `filing_no`, which is wrapped inside two structs. Both the `export` struct and the `license` struct inside it are nullable, i.e. not all shipments are international and not all export shipments need a license.
```

Comment:
Compared to the previous query there is almost 2x increase in CPU and more than 2x increase in wall time. I think readers will be surprised. Do we attribute the different to reading null flags? If so, would it make sense to include an equivalent of the previous query with the only difference being that columns are nullable?

Reply:
Added a commit explaining.